File: packages/cli/src/commands/fetch.ts
import process from 'node:process'
import { defineCommand } from 'citty'
import { ensureCheckout as defaultEnsureCheckout } from './ensure-checkout.js'

export interface RunFetchOptions {
  specs: string[]
  projectDir: string
  quiet?: boolean
}

export interface RunFetchDeps {
  ensureCheckout?: typeof defaultEnsureCheckout
  log?: (msg: string) => void
  error?: (msg: string) => void
  exit?: (code: number) => void
}

/**
 * Implementation of `ask fetch <spec...>` (ported from opensrc's
 * `opensrc fetch`, vercel-labs/opensrc#53). Warms the checkout cache for
 * each spec via the shared `ensureCheckout` helper WITHOUT printing
 * paths — the counterpart to `ask src`/`ask docs`, which print where the
 * source lives. Useful for prefetching in setup scripts and for warming
 * the cache that `ask add`'s offline-first docs-path prompt reads from.
 *
 * Per-spec failures are reported and the remaining specs still run;
 * the process exits non-zero at the end if any spec failed.
 */
export async function runFetch(options: RunFetchOptions, deps: RunFetchDeps = {}): Promise<void> {
  const ensureCheckout = deps.ensureCheckout ?? defaultEnsureCheckout
  const log = deps.log ?? ((msg: string) => process.stdout.write(`${msg}\n`))
  const error = deps.error ?? ((msg: string) => process.stderr.write(`${msg}\n`))
  const exit = deps.exit ?? ((code: number) => process.exit(code))

  let fetched = 0
  let cached = 0
  let hadErrors = false

  for (const spec of options.specs) {
    try {
      const result = await ensureCheckout({ spec, projectDir: options.projectDir })
      // Avoid `spec@ref` duplication when the user already pinned the
      // ref in the spec itself (e.g. `github:owner/repo@v1.2.3`).
      const display = spec.endsWith(`@${result.ref}`) ? spec : `${spec}@${result.ref}`
      if (result.fromCache) {
        cached += 1
        if (!options.quiet)
          log(`  ✓ ${display} already cached (${result.checkoutDir})`)
      }
      else {
        fetched += 1
        if (!options.quiet)
          log(`  ✓ Fetched ${display} (${result.checkoutDir})`)
      }
    }
    catch (err) {
      hadErrors = true
      error(`  ✗ ${spec}: ${err instanceof Error ? err.message : String(err)}`)
    }
  }

  if (!options.quiet) {
    const parts: string[] = []
    if (fetched > 0)
      parts.push(`${fetched} fetched`)
    if (cached > 0)
      parts.push(`${cached} already cached`)
    if (parts.length > 0)
      log(`\n${parts.join(', ')}`)
  }

  if (hadErrors)
    exit(1)
}

/**
 * Citty command surface for `ask fetch`. Wires the citty argument parser
 * to `runFetch` and uses real stdout/stderr/process.exit. Tests should
 * call `runFetch` directly with mocked deps.
 */
export const fetchCmd = defineCommand({
  meta: {
    name: 'fetch',
    description: 'Warm the source cache for one or more specs without printing paths',
  },
  args: {
    spec: {
      type: 'positional',
      description: 'Library spec(s) (e.g. react, npm:react@18.2.0, github:facebook/react@v18.2.0)',
      required: true,
    },
    quiet: {
      type: 'boolean',
      alias: 'q',
      description: 'Suppress progress output (errors still print to stderr)',
    },
  },
  async run({ args }) {
    // citty binds the first positional to `args.spec` and keeps ALL
    // positionals (including the first) in `args._`.
    const rest = (args._ ?? []).filter(s => s.length > 0)
    const specs = rest.length > 0 ? rest : [args.spec]
    await runFetch({
      specs,
      projectDir: process.cwd(),
      quiet: Boolean(args.quiet),
    })
  },
})

File: crates/ask/src/commands/fetch.rs
//! `ask fetch <spec...>` — warm the source cache for one or more specs without
//! printing paths. Rust port of `commands/fetch.ts` (itself ported from
//! opensrc's `opensrc fetch`, vercel-labs/opensrc#53).
//!
//! Per-spec failures are reported and the remaining specs still run; the CLI
//! exits non-zero at the end if any spec failed. `run_fetch` returns a
//! structured report so the streaming/exit behaviour is unit-testable offline.

use std::path::PathBuf;

use crate::commands::ensure_checkout::{
    ensure_checkout, EnsureCheckoutDeps, EnsureCheckoutOptions,
};
use crate::http::HttpClient;

/// Options for [`run_fetch`].
#[derive(Debug, Clone)]
pub struct RunFetchOptions {
    pub specs: Vec<String>,
    pub project_dir: PathBuf,
    pub quiet: bool,
}

/// Outcome of [`run_fetch`]. The CLI writes `stdout` lines to stdout, `stderr`
/// lines to stderr, and exits non-zero when `had_errors`.
#[derive(Debug, Default, PartialEq, Eq)]
pub struct FetchReport {
    pub stdout: Vec<String>,
    pub stderr: Vec<String>,
    pub had_errors: bool,
}

/// Warm the cache for each spec via the shared checkout helper.
pub fn run_fetch(client: &dyn HttpClient, options: &RunFetchOptions) -> FetchReport {
    run_fetch_with(client, options, &EnsureCheckoutDeps::default())
}

/// [`run_fetch`] with injectable checkout deps.
pub fn run_fetch_with(
    client: &dyn HttpClient,
    options: &RunFetchOptions,
    deps: &EnsureCheckoutDeps,
) -> FetchReport {
    let mut report = FetchReport::default();
    let mut fetched = 0usize;
    let mut cached = 0usize;

    for spec in &options.specs {
        match ensure_checkout(
            client,
            &EnsureCheckoutOptions {
                spec: spec.clone(),
                project_dir: options.project_dir.clone(),
                no_fetch: false,
            },
            deps,
        ) {
            Ok(result) => {
                // Avoid `spec@ref` duplication when the user already pinned the
                // ref in the spec itself (e.g. `github:owner/repo@v1.2.3`).
                let suffix = format!("@{}", result.reference);
                let display = if spec.ends_with(&suffix) {
                    spec.clone()
                } else {
                    format!("{spec}{suffix}")
                };
                let path = result.checkout_dir.to_string_lossy();
                if result.from_cache {
                    cached += 1;
                    if !options.quiet {
                        report
                            .stdout
                            .push(format!("  ✓ {display} already cached ({path})"));
                    }
                } else {
                    fetched += 1;
                    if !options.quiet {
                        report
                            .stdout
                            .push(format!("  ✓ Fetched {display} ({path})"));
                    }
                }
            }
            Err(err) => {
                report.had_errors = true;
                report.stderr.push(format!("  ✗ {spec}: {err}"));
            }
        }
    }

    if !options.quiet {
        let mut parts: Vec<String> = Vec::new();
        if fetched > 0 {
            parts.push(format!("{fetched} fetched"));
        }
        if cached > 0 {
            parts.push(format!("{cached} already cached"));
        }
        if !parts.is_empty() {
            report.stdout.push(format!("\n{}", parts.join(", ")));
        }
    }

    report
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::http::mock::MockClient;
    use crate::store::github_store_path;

    fn deps_home(home: &std::path::Path) -> EnsureCheckoutDeps<'static> {
        EnsureCheckoutDeps {
            ask_home: Some(home.to_path_buf()),
            fetcher: None,
        }
    }

    #[test]
    fn reports_already_cached_without_duplicating_ref() {
        let home = tempfile::tempdir().unwrap();
        let proj = tempfile::tempdir().unwrap();
        let dir = github_store_path(home.path(), "github.com", "owner", "repo", "v1.2.3").unwrap();
        std::fs::create_dir_all(&dir).unwrap();
        let report = run_fetch_with(
            &MockClient::new(),
            &RunFetchOptions {
                specs: vec!["github:owner/repo@v1.2.3".into()],
                project_dir: proj.path().to_path_buf(),
                quiet: false,
            },
            &deps_home(home.path()),
        );
        assert!(!report.had_errors);
        // The spec already ends with @v1.2.3 — no `@v1.2.3@v1.2.3` duplication.
        assert!(report.stdout[0].contains("github:owner/repo@v1.2.3 already cached"));
        assert!(!report.stdout[0].contains("v1.2.3@v1.2.3"));
        assert!(report.stdout.last().unwrap().contains("1 already cached"));
    }

    #[test]
    fn per_spec_failure_sets_error_and_continues() {
        let home = tempfile::tempdir().unwrap();
        let proj = tempfile::tempdir().unwrap();
        // First spec resolves (cache hit); second is an unsupported ecosystem.
        let dir = github_store_path(home.path(), "github.com", "owner", "repo", "v1.0.0").unwrap();
        std::fs::create_dir_all(&dir).unwrap();
        let report = run_fetch_with(
            &MockClient::new(),
            &RunFetchOptions {
                specs: vec![
                    "github:owner/repo@v1.0.0".into(),
                    "cargo:serde".into(), // no resolver for cargo
                ],
                project_dir: proj.path().to_path_buf(),
                quiet: false,
            },
            &deps_home(home.path()),
        );
        assert!(report.had_errors);
        assert_eq!(report.stderr.len(), 1);
        assert!(report.stderr[0].contains("cargo:serde"));
        // The good spec still succeeded.
        assert!(report.stdout.iter().any(|l| l.contains("already cached")));
    }

    #[test]
    fn quiet_suppresses_stdout_but_keeps_errors() {
        let home = tempfile::tempdir().unwrap();
        let proj = tempfile::tempdir().unwrap();
        let report = run_fetch_with(
            &MockClient::new(),
            &RunFetchOptions {
                specs: vec!["cargo:serde".into()],
                project_dir: proj.path().to_path_buf(),
                quiet: true,
            },
            &deps_home(home.path()),
        );
        assert!(report.had_errors);
        assert!(report.stdout.is_empty());
        assert_eq!(report.stderr.len(), 1);
    }
}
